File: src/bezier/Bezier.java
package bezier;

import javafx.scene.control.Alert;

import java.util.ArrayList;

import static java.util.stream.Collectors.toList;
import static utils.Config.*;
import static utils.UnitConverter.*;

public class Bezier {

    public static ArrayList<Point> generateAll(ArrayList<Point> controlPoints) {
        ArrayList<Point> pathPureXY = generatePureXY(controlPoints);
        ArrayList<Double> times = trapezoidalTimes(pathPureXY, controlPoints);
        ArrayList<Point> path = generateWithVel(controlPoints, times);
        motion(path);
        return path;
    }

    public static ArrayList<Point> generatePureXY(ArrayList<Point> controlPoints) {
        ArrayList<Point> path = new ArrayList<>();
        ArrayList<Point> throughPoints = (ArrayList<Point>) controlPoints.stream()
                .filter(Point::isIntercept)
                .collect(toList());
        for (int j = 0; j < throughPoints.size() - 1; j++) {
            for (double t = 0.0; t <= 1; t += 1. / 299.) { //some large number of saves.points
                double sumX = 0, sumY = 0;
                double T = 1 - t;
                int startPoint = controlPoints.indexOf(throughPoints.get(j));
                int endPoint = controlPoints.indexOf(throughPoints.get(j + 1));
                for (int i = 0; i <= endPoint - startPoint; i++) {
                    int I = (endPoint - startPoint) - i;
                    sumX += polynomialCoeff(endPoint - startPoint, i) * Math.pow(T, i) * Math.pow(t, I) * controlPoints.get(I + startPoint).getX();
                    sumY += polynomialCoeff(endPoint - startPoint, i) * Math.pow(T, i) * Math.pow(t, I) * controlPoints.get(I + startPoint).getY();
                }
                path.add(new Point(sumX, sumY));
                if (t == 0)
                    path.get(path.size() - 1).setIntercept(true); //start and end of each segment is at same point as an intercept
            }
        }
        if (path.isEmpty()) return new ArrayList<>();
        path.get(0).setDistance(0);
        path.get(path.size() - 1).setIntercept(true);
        for (int i = 0; i < path.size(); i++) {
            Point p = path.get(i);
            if (i != path.size() - 1) {
                p.setHeadingTo(path.get(i + 1));
            }
            if (i != 0) {
                p.setDistance(path.get(i - 1).getDistance() + p.distanceTo(path.get(i - 1)));
            }
        }
        path.get(path.size() - 1).setHeading(path.get(path.size() - 2).getHeading());
        return path;
    }

    public static ArrayList<Double> trapezoidalTimes(ArrayList<Point> pathXY, ArrayList<Point> controlPoints) {
        if (pathXY.isEmpty()) return new ArrayList<>();
        ArrayList<Double> times = new ArrayList<>();
        ArrayList<Point> throughPoints = (ArrayList<Point>) pathXY.stream()
                .filter(Point::isIntercept)
                .collect(toList());
        ArrayList<Point> oldThroughPoints = (ArrayList<Point>) controlPoints.stream()
                .filter(Point::isIntercept)
                .collect(toList());
        for (int j = 0; j < throughPoints.size() - 1; j++) {
            double lengthOfCurve = throughPoints.get(j + 1).getDistance() - throughPoints.get(j).getDistance();
            lengthOfCurve = inchesToFeet(lengthOfCurve);
            double velInitial = oldThroughPoints.get(j).getTargetVelocity();
            double velFinal = oldThroughPoints.get(j + 1).getTargetVelocity();
            double velMax = oldThroughPoints.get(j + 1).isOverrideMaxVel()
                    ? oldThroughPoints.get(j + 1).getTargetVelocity()
                    : maxVel();

            if (velMax == 0) {
                String errorMsg = "with a max speed of 0, you'll never get anywhere!";
                Alert alert = new Alert(Alert.AlertType.ERROR);
                alert.setHeaderText("Impossible Physics");
                alert.setContentText(errorMsg);
                alert.showAndWait();
                throw new IllegalStateException(errorMsg);
            }

            //if accel and deccel take same time, calculations are much easier
            double velInitialAndFinal = velInitial;
            double timeAccelInitToFinal = 0;
            if (velInitial != velFinal) {
                timeAccelInitToFinal = Math.abs(velInitial - velFinal) / maxAccel();
                double distToEqualizeVels = timeAccelInitToFinal * (velInitial + velFinal) / 2.;
                velInitialAndFinal = Math.min(velInitial, velFinal) + Math.abs(velInitial - velFinal);

                if (distToEqualizeVels > lengthOfCurve) {
                    String errorMsg = "distance traveled while accelerating from initial to final velocity ("
                            + distToEqualizeVels + ") is greater than total distance of path (" + lengthOfCurve + ")";
                    Alert alert = new Alert(Alert.AlertType.ERROR);
                    alert.setHeaderText("Impossible Physics");
                    alert.setContentText(errorMsg);
                    alert.showAndWait();
                    throw new IllegalStateException(errorMsg);
                }
                lengthOfCurve -= distToEqualizeVels;
            }

            //calculate max vel that is reachable physically
            double timeAccelTriangle = quadratic(0.5 * maxAccel(), velInitial, -lengthOfCurve / 2); //x = v0t + 1/2at^2
            double velMaxReachable = Math.min(velInitialAndFinal + maxAccel() * timeAccelTriangle, velMax);

            double timeAccel = (velMaxReachable - velInitialAndFinal) / maxAccel(),
                    timeDeccel = (velInitialAndFinal - velMaxReachable) / -maxAccel();
            double timeConst = (lengthOfCurve - ((timeAccel * (velInitialAndFinal + velMaxReachable) / 2) + (timeDeccel * (velInitialAndFinal + velMaxReachable) / 2))) / velMaxReachable;
            times.add(timeAccelInitToFinal + timeAccel + timeConst + timeDeccel);

//            //if triangle, quadratic; if trapezoid vel. the smaller is the always the right one //TODO "ALWAYS"?
//            double timeAccelTrapezoid = (velMax - velInitial) / maxAccel(); //t = deltaV / a
//            if (timeAccelTrapezoid < timeAccelTriangle) { //is a trapezoid
//                double timeConst = (lengthOfCurve - 2 * timeAccelTrapezoid * (velInitial + velMax)) / maxAccel();
//                times.add(timeAccelTrapezoid + timeConst + timeAccelTrapezoid);
//            } else { //is a triangle
//                times.add(timeAccelTriangle + timeAccelTriangle);
//            }
        }
        return times;
    }

    public static ArrayList<Point> generateWithVel(ArrayList<Point> controlPoints, ArrayList<Double> times) {
        ArrayList<Point> path = new ArrayList<>();
        ArrayList<Point> throughPoints = (ArrayList<Point>) controlPoints.stream()
                .filter(Point::isIntercept)
                .collect(toList());

        for (int j = 0; j < throughPoints.size() - 1; j++) {
            double startVel = throughPoints.get(j).getTargetVelocity(),
                    endVel = throughPoints.get(j + 1).getTargetVelocity(); //problem is that is is being changed in prior runnings of this method
            double curMaxVel = maxVel();
            if (controlPoints.get(j + 1).isOverrideMaxVel())
                curMaxVel = Math.max(startVel, endVel);
            double time = times.get(j);
            double precision = Math.ceil(time / timeStep()) + 1;
            for (int fakeT = 0; fakeT <= precision; fakeT++) {
                double t = fakeT / precision;
                double sumX = 0, sumY = 0;
                double T = 1 - t;
                int startPoint = controlPoints.indexOf(throughPoints.get(j));
                int endPoint = controlPoints.indexOf(throughPoints.get(j + 1));
                for (int i = 0; i <= endPoint - startPoint; i++) {
                    int I = (endPoint - startPoint) - i;
                    sumX += polynomialCoeff(endPoint - startPoint, i) * Math.pow(T, i) * Math.pow(t, I) * controlPoints.get(I + startPoint).getX();
                    sumY += polynomialCoeff(endPoint - startPoint, i) * Math.pow(T, i) * Math.pow(t, I) * controlPoints.get(I + startPoint).getY();
                }
                path.add(new Point(sumX, sumY));
                //trapezoidal velocities
                double up = Math.min(maxAccel() * t * time + startVel, curMaxVel),
                        down = Math.min(-maxAccel() * (t * time - time) + endVel, curMaxVel);
                path.get(path.size() - 1).setTargetVelocity(Math.min(up, down));
                path.get(path.size() - 1).setTime(time * t);
            }
            if (throughPoints.get(j + 1).isReverse())
                path.stream().skip((long) (path.size() - precision)).forEach(Point::reverse);
            if (j != throughPoints.size() - 2 && !path.isEmpty()) path.remove(path.size() - 1);
        }
        if (path.isEmpty()) return new ArrayList<>();
        path.get(0).setDistance(0);
        for (int i = 0; i < path.size(); i++) {
            Point p = path.get(i);
            if (i != path.size() - 1) {
                p.setHeadingTo(path.get(i + 1));
            }
            if (i != 0) {
                p.setDistance(path.get(i - 1).getDistance() + p.distanceTo(path.get(i - 1)));
            }
        }
        path.get(path.size() - 1).setHeading(path.get(path.size() - 2).getHeading());
        return path;
    }

    /**
     * <p>Calculates velocity and position of each point in the path</p>
     * <p>turning left, equation is:</p>
     * <p>Vr = w * (R + l/2)</p>
     * <p>Vl = w * (R - l/2)</p>
     *
     * @param path contains x,y,theta coordinates of each point, and the target velocity to travel at (as if going in a line)
     */
    public static void motion(ArrayList<Point> path) {
        if (path.size() == 0) return;
        double halfWidth = width() / 2;
        for (int i = 0; i < path.size(); i++) {
            int iAdjusted = i;
            if (i + 2 > path.size() - 1)
                iAdjusted = path.size() - 3;
            double circleRadius = radiusOfCircle(path.get(iAdjusted), path.get(iAdjusted + 1), path.get(iAdjusted + 2));
            double leftVel, rightVel;
            if (Double.isInfinite(circleRadius)) { //linear path
                leftVel = feetToInches(path.get(i).getTargetVelocity());
                rightVel = feetToInches(path.get(i).getTargetVelocity());
            } else {
                double angularVel = feetToInches(path.get(i).getTargetVelocity()) / circleRadius;
                if (path.get(iAdjusted + 2).getHeadingCartesian() < path.get(iAdjusted).getHeadingCartesian()) { // turning left
                    leftVel = angularVel * (circleRadius - halfWidth);
                    rightVel = angularVel * (circleRadius + halfWidth);
                } else { //turning right
                    leftVel = angularVel * (circleRadius + halfWidth);
                    rightVel = angularVel * (circleRadius - halfWidth);
                }
            }
            leftVel = linearToRotational(leftVel);
            rightVel = linearToRotational(rightVel);
            path.get(i).setVels(leftVel, rightVel);
            if (i == 0) {
                path.get(i).setPos(rotationalToLinear(0), rotationalToLinear(0));
            } else path.get(i).advancePos(path.get(i - 1).getLeftPos(), path.get(i - 1).getRightPos());
        }
    }

    /**
     * <p>calculates the radius of the circle which the 3 saves.points lie on</p>
     * <p>r = a*b*c/4*area</p>
     *
     * @param circlePoints 3 saves.points from which a circle is extrapolated
     * @return radius of the circle in same unit as represented in the circlePoints in inches
     */
    private static double radiusOfCircle(Point... circlePoints) {
        assert circlePoints.length == 3;
        double a = circlePoints[0].distanceTo(circlePoints[1]),
                b = circlePoints[1].distanceTo(circlePoints[2]),
                c = circlePoints[2].distanceTo(circlePoints[0]);
        double s = (a + b + c) / 2;
        a = Math.min(s, a);     //fix rounding error
        b = Math.min(s, b);     //if any were > s, NaN results
        c = Math.min(s, c);
        double area = Math.sqrt(s * (s - a) * (s - b) * (s - c));
        return (a * b * c) / (4 * area);
    }

    private static double quadratic(double a, double b, double c) {
        return Math.max(
                (-b + Math.sqrt(b * b - 4 * a * c)) / (2 * a),
                (-b + Math.sqrt(b * b - 4 * a * c)) / (2 * a));
    }

    private static int polynomialCoeff(int line, int n) {
        int result = 1;
        for (int i = 0; i < n; i++) {
            result *= line - i;
            result /= i + 1;
        }
        return result;
    }
}

File: src/main/java/bezier/Bezier.kt
package bezier

import bezier.units.*
import bezier.units.derived.*
import java.util.*
import kotlin.math.pow

object Bezier {
    @JvmStatic
    fun generateSpline(controlPoints: ArrayList<Point>,
                       maxVel: InchesPerSecond,
                       maxAccel: InchesPerSecondSquared,
                       timeStep: Seconds,
                       width: Inches
    ): ArrayList<Point> {
        val pathXYCoords = generatePureXY(controlPoints)
        val times = trapezoidalTimes(pathXYCoords, maxVel, maxAccel)
        val path = generateWithVel(controlPoints, times, maxVel, maxAccel, timeStep)
        return motion(path, width)
    }

    /**
     * generates the x and y coordinates and headings of each point on the Bezier curve defined by [controlPoints]
     * marks points with the same x and y coordinates as in [controlPoints] as interceptPoints
     */
    @JvmStatic
    fun generatePureXY(controlPoints: ArrayList<Point>): ArrayList<Point> {
        val path = ArrayList<Point>()
        val throughPoints = controlPoints.filter { it.isIntercept }
        val precision = 299

        for (j in 0 until throughPoints.size - 1) {
            for (fakeT in 0..precision) {
                val t = fakeT.toDouble() / precision
                val T = 1.0 - t
                var sumX = 0.inches()
                var sumY = 0.inches()
                val startPoint = controlPoints.indexOf(throughPoints[j])
                val endPoint = controlPoints.indexOf(throughPoints[j + 1])
                for (i in 0..endPoint - startPoint) {
                    val I = endPoint - startPoint - i
                    sumX += controlPoints[I + startPoint].x * polynomialCoeff(endPoint - startPoint, i) * T.pow(i) * t.pow(I)
                    sumY += controlPoints[I + startPoint].y * polynomialCoeff(endPoint - startPoint, i) * T.pow(i) * t.pow(I)
                }
                val currentPoint = Point(sumX, sumY)
                if (t == 0.0) {
                    if (j != 0) continue // don't add duplicate points, since start and end of consecutive paths are the same
                    currentPoint.isIntercept = true // start and end of each segment is at same point as an intercept
                    currentPoint.distance = 0.inches()
                    currentPoint.targetVelocity = throughPoints[0].targetVelocity
                } else {
                    if (t == 1.0) {
                        currentPoint.isIntercept = true
                        currentPoint.targetVelocity = throughPoints[j].targetVelocity
                    }
                    currentPoint.distance = path.last().distance + path.last().distanceTo(currentPoint)
                    path.last().setHeadingTo(currentPoint)
                    path.last().setLeftAndRightPositions()
                }
                path += currentPoint
            }
        }
        if (path.isEmpty()) return ArrayList()

        path.last().heading = path[path.size - 2].heading
        path.last().setLeftAndRightPositions()
        path.last().isIntercept = true

        for (point in path.withIndex()) {
            if (point.index == 0) point.value.distance = 0.inches()
            else {
                point.value.leftPoint?.distance = path[point.index - 1].leftPoint!!.distance + path[point.index - 1].leftPoint!!.distanceTo(point.value.leftPoint!!)
                point.value.rightPoint?.distance = path[point.index - 1].rightPoint!!.distance + path[point.index - 1].rightPoint!!.distanceTo(point.value.rightPoint!!)
            }
        }

        return path
    }

    @JvmStatic
    fun trapezoidalTimes(pathXY: ArrayList<Point>,
//                         controlPoints: ArrayList<Point>,
                         maxVel: InchesPerSecond,
                         maxAccel: InchesPerSecondSquared
    ): ArrayList<Seconds> {
        if (pathXY.isEmpty()) return ArrayList()

        val times = ArrayList<Seconds>()
        val throughPoints = pathXY.filter { it.isIntercept }
//        val oldThroughPoints = controlPoints.filter { it.isIntercept }

        for (j in 0 until throughPoints.size - 1) {
            var lengthOfCurve = throughPoints[j + 1].distance - throughPoints[j].distance

            val velInitial: InchesPerSecond = throughPoints[j].targetVelocity
            val velFinal: InchesPerSecond = throughPoints[j + 1].targetVelocity
            val velMax: InchesPerSecond =
                    if (throughPoints[j + 1].isOverrideMaxVel) throughPoints[j + 1].targetVelocity
                    else maxVel.inchesPerSecond()

            if (velMax.value == 0.0)
                throw IllegalStateException("with a max speed of 0, you'll never get anywhere!")

            //if accel and deccel take same amount of time, calculations are much easier
            val velInitialAndFinal: InchesPerSecond = maxOf(velInitial, velFinal)
            var timeAccelInitToFinal = 0.seconds()

            if (velInitial != velFinal) {
                timeAccelInitToFinal = (velInitial - velFinal).abs() / maxAccel.inchesPerSecondSquared()
                val distToEqualizeVels: Inches = (velInitial + velFinal) / 2.0 * timeAccelInitToFinal

                if (distToEqualizeVels > lengthOfCurve)
                    throw IllegalStateException("distance traveled while accelerating from initial to final velocity (${distToEqualizeVels.value}) " +
                            "is greater than total distance of path (${lengthOfCurve.value})"
                    )
                lengthOfCurve -= distToEqualizeVels
            }

            //calculate max vel that is reachable physically
            val timeAccelHalfTriangle: Seconds = quadratic(maxAccel.inchesPerSecondSquared() * 0.5, velInitialAndFinal, -lengthOfCurve / 2)
            val maxVelReachable: InchesPerSecond = minOf(velMax, velInitialAndFinal + (maxAccel * timeAccelHalfTriangle).inchesPerSecond())

            val timeAccel = (maxVelReachable - velInitialAndFinal) / maxAccel.inchesPerSecondSquared()
            val timeDeccel = (velInitialAndFinal - maxVelReachable) / -maxAccel.inchesPerSecondSquared()
            val timeConst = (lengthOfCurve - (((velInitialAndFinal + maxVelReachable) * timeAccel) / 2 + ((velInitialAndFinal + maxVelReachable)) * timeDeccel / 2)) / maxVelReachable
            times += timeAccelInitToFinal + timeAccel + timeConst + timeDeccel
        }
        return times
    }

    @JvmStatic
    fun <L : Length<L>> trapezoidalTime(startPos: L,
                                        endPos: L,
                                        maxVel: LinearVelocity<L, Seconds>,
                                        maxAccel: Acceleration<L, Seconds>
    ): Seconds {
        val length = endPos - startPos
        val initialVel = maxVel.createNew(0.0)

        val timeAccelTriangle: Seconds = quadratic(maxAccel / 2, maxVel.createNew(0.0), -length / 2)
        val maxVelReachable: LinearVelocity<L, Seconds> = minOf(maxVel, initialVel + (maxAccel * timeAccelTriangle))

        val timeAccel: Seconds = (maxVelReachable - initialVel) / maxAccel
        val timeDeccel: Seconds = (initialVel - maxVelReachable) / -maxAccel
        val timeConst: Seconds = (length - ((initialVel + maxVelReachable) * timeAccel / 2 + (initialVel + maxVelReachable) * timeDeccel / 2)) / maxVelReachable

        return timeAccel + timeConst + timeDeccel
    }

    @JvmStatic
    fun generateWithVel(controlPoints: ArrayList<Point>,
                        times: ArrayList<Seconds>,
                        maxVel: InchesPerSecond,
                        maxAccel: InchesPerSecondSquared,
                        timeStep: Seconds
    ): ArrayList<Point> {
        if (times.isEmpty()) return ArrayList()

        val path = ArrayList<Point>()
        val throughPoints = controlPoints.filter { it.isIntercept }

        var prevEnd = 0 // TODO make this less terrible
        for (j in 0 until throughPoints.size - 1) {
            val startVel: InchesPerSecond = throughPoints[j].targetVelocity
            val endVel: InchesPerSecond = throughPoints[j + 1].targetVelocity
            val curMaxVel: InchesPerSecond =
                    if (controlPoints[j + 1].isOverrideMaxVel) maxOf(startVel, endVel)
                    else maxVel.inchesPerSecond()

            val time: Seconds = times[j]
            val precision: Int = Math.floor(time / timeStep).toInt()
            for (fakeT in 0..precision) {
                val t = fakeT / precision.toDouble() // 0 to 1 with precision iterations
                var sumX = Inches(0.0)
                var sumY = Inches(0.0)
                val T = 1 - t
                val startPoint = controlPoints.indexOf(throughPoints[j])
                val endPoint = controlPoints.indexOf(throughPoints[j + 1])
                for (i in 0..endPoint - startPoint) {
                    val I = endPoint - startPoint - i
                    sumX += controlPoints[I + startPoint].x * polynomialCoeff(endPoint - startPoint, i).toDouble() * Math.pow(T, i.toDouble()) * Math.pow(t, I.toDouble())
                    sumY += controlPoints[I + startPoint].y * polynomialCoeff(endPoint - startPoint, i).toDouble() * Math.pow(T, i.toDouble()) * Math.pow(t, I.toDouble())
                }
                path += Point(sumX, sumY)

                //trapezoidal velocities
                val up: InchesPerSecond = minOf(curMaxVel, startVel + (maxAccel * (time * t)).inchesPerSecond())
                val down: InchesPerSecond = minOf(curMaxVel, endVel - (maxAccel * (time * t - time)).inchesPerSecond())
                path.last().targetVelocity = minOf(up, down)
                path.last().time = time * t + path[prevEnd].time
                //                if (j != 0) path.get(path.size() - 1).setTime(denominOfator*t + path.get(prevEnd).getDenominOfator());            //TODO where did this come from? should it be here?
                //                else path.get(path.size() - 1).setTime(denominOfator * t);          //TODO where did this come from? should it be here?
//                if (j != 0)
//                    path.last().time =
            }
            if (throughPoints[j + 1].isReverse)
                path.stream().skip((path.size - precision).toLong()).forEach(Point::reverse)

            if (j != throughPoints.size - 2 && !path.isEmpty()) path.removeAt(path.size - 1)
            prevEnd = path.size - 1
        }
        if (path.size < 2) return ArrayList()

        path[0].distance = Inches(0.0)
        for (i in path.indices) {
            val p = path[i]
            if (i != path.size - 1) {
                p.setHeadingTo(path[i + 1])
            }
            if (i != 0) {
                p.distance = path[i - 1].distance + p.distanceTo(path[i - 1])
            }
        }
        path.last().heading = path[path.size - 2].heading
        return path
    }

    /**
     * for linear motion of elevator
     */
    @JvmStatic
    fun <L : Length<L>> generateVelProfile(time: Seconds,
                                           timeStep: Seconds,
                                           maxVel: LinearVelocity<L, Seconds>,
                                           maxAccel: Acceleration<L, Seconds>
    ): Array<LinearVelocity<L, Seconds>> {
        val precision: Int = Math.floor(time / timeStep).toInt()
        return Array(precision) { i ->
            val t = i / precision.toDouble()
            val up: LinearVelocity<L, Seconds> = minOf(maxVel, (maxAccel * t * time))
            val down: LinearVelocity<L, Seconds> = minOf(maxVel, (-maxAccel * (time * t - time)))
            minOf(up, down)
        }
    }

    /**
     * Calculates velocity and position of each point in the path
     * turning left, equation is:
     *
     * Vr = w * (R + l/2)
     *
     * Vl = w * (R - l/2)
     *
     * @param path contains x,y,theta coordinates of each point, and the target velocity to travel at (as if going in a line)
     */
    @JvmStatic
    fun motion(path: ArrayList<Point>, width: Inches): ArrayList<Point> {
        if (path.isEmpty()) return arrayListOf()

        val halfWidth: Inches = (width / 2)
        for (i in path.indices) {
            val iAdjusted = if (i + 2 > path.size - 1) path.size - 3 else i
            val circleRadius = radiusOfCircle(path[iAdjusted], path[iAdjusted + 1], path[iAdjusted + 2])

            var leftWheelVel: InchesPerSecond
            var rightWheelVel: InchesPerSecond

            //TODO more cases for when turning radius is less that robot radius
            if (circleRadius > 5_000) { //straight line path (some are not infinite, just very big)
                leftWheelVel = path[i].targetVelocity
                rightWheelVel = path[i].targetVelocity
            } /*else if (circleRadius < width) { //maybe should be halfWidth
                TODO("more math needed for turns that are inside frame perimeter")
            } */ else {
                val angularVel = RadiansPerSecond(path[i].targetVelocity, circleRadius)

                if (path[iAdjusted].heading isLeftTowards path[iAdjusted + 2].heading) { // turning left
                    leftWheelVel = angularVel * (circleRadius - halfWidth)
                    rightWheelVel = angularVel * (circleRadius + halfWidth)
                } else { //turning right
                    leftWheelVel = angularVel * (circleRadius + halfWidth)
                    rightWheelVel = angularVel * (circleRadius - halfWidth)
                }
            }

            path[i].setVels(leftWheelVel, rightWheelVel)
            if (i == 0)
                path[i].setPos(0.inches(), 0.inches())
            else
                path[i].advancePos(path[i - 1].leftPos, path[i - 1].rightPos)
        }
        return path
    }

    /**
     * velocity profile for linear motion (elevator)
     */
    fun <L : Length<L>> generateLinear(startPos: L, endPos: L): Array<LinearVelocity<L, Seconds>> {
        val maxVel = startPos.createNew(10.0) / 1.seconds()
        val maxAccel = startPos.createNew(10.0) / 1.seconds() / 1.seconds()

        val time = trapezoidalTime(startPos, endPos, maxVel, maxAccel)
        return generateVelProfile(time, 0.02.seconds(), maxVel, maxAccel)
    }

    //I was not source of equations in methods below this comment

    /**
     *
     * calculates the radius of the circle which the 3 points lie on
     *
     * r = a*b*c/4*area
     *
     * @param circlePoints 3 points from which a circle is extrapolated
     * @return radius of the circle in same unit as represented in the circlePoints in inches
     */
    private fun radiusOfCircle(vararg circlePoints: Point): Inches {
        if (circlePoints.size != 3) throw InputMismatchException("radiusOfCircle needs exactly THREE points")

        var a = circlePoints[0].distanceTo(circlePoints[1])
        var b = circlePoints[1].distanceTo(circlePoints[2])
        var c = circlePoints[2].distanceTo(circlePoints[0])
        val s = (a + b + c) / 2
        a = minOf(s, a)
        b = minOf(s, b)
        c = minOf(s, c)
        val area = Math.sqrt(s.value * (s - a).value * (s - b).value * (s - c).value)
        return (a * b.value * c.value) / (4 * area)
    }

    /**
     * quadratic equation to solve 0=ax^2+bx+c
     * solves for positive values
     */
    private fun quadratic(a: Double, b: Double, c: Double): Double {
        return Math.max(
                (-b + Math.sqrt(b * b - 4.0 * a * c)) / (2 * a),
                (-b - Math.sqrt(b * b - 4.0 * a * c)) / (2 * a))
    }

    /**
     * type safe quadratic formula
     */
    private fun <L : Length<L>, T : Time<T>> quadratic(a: Acceleration<L, T>, b: LinearVelocity<L, T>, c: L): T {
        return b.createNewTime(quadratic(a.value, b.value, c.value))
    }

    /**
     * solves for number in Pascal's triangle on a [line] [n] indices over
     */
    private fun polynomialCoeff(line: Int, n: Int): Int {
        var result = 1
        for (i in 0 until n) {
            result *= line - i
            result /= i + 1
        }
        return result
    }
}
